Mathlib/NumberTheory/LSeries/Convolution.lean

```diff
@@ -83,52 +83,56 @@ lemma convolution_map_zero {R : Type*} [Semiring R] (f g : ℕ → R) : (f ⍟ g
 ### Multiplication of L-series
 -/
 
+/-- We give an expression of the `LSeries.term` of the convolution of two functions
+in terms of a sum over `Nat.divisorsAntidiagonal`. -/
+lemma term_convolution' (f g : ℕ → ℂ) (s : ℂ) (n : ℕ) :
+    term (f ⍟ g) s n = ∑ p in n.divisorsAntidiagonal, term f s p.1 * term g s p.2 := by
+  rcases eq_or_ne n 0 with rfl | hn
+  · simp only [term, ↓reduceIte, Nat.divisorsAntidiagonal_zero, mul_ite, mul_zero, ite_mul,
+      zero_mul, Finset.sum_empty]
+  -- now `n ≠ 0`
+  rw [term_of_ne_zero hn, convolution_def, Finset.sum_div]
+  refine Finset.sum_congr rfl fun p hp ↦ ?_
+  obtain ⟨hp, hn₀⟩ := Nat.mem_divisorsAntidiagonal.mp hp
+  have ⟨hp₁, hp₂⟩ := mul_ne_zero_iff.mp <| hp.symm ▸ hn₀
+  rw [term_of_ne_zero hp₁ f s, term_of_ne_zero hp₂ g s, mul_comm_div, div_div,
+    ← mul_div_assoc, ← natCast_mul_natCast_cpow, ← Nat.cast_mul, mul_comm p.2, hp]
+
 open Set Nat in
-/-- We give an expression of the `LSeries.term` of the convolution of two functions. -/
+/-- We give an expression of the `LSeries.term` of the convolution of two functions
+in terms of an a priori infinte sum over all pairs `(k, m)` with `k * m = n`
+(the set we sum over is infinite when `n = 0`). This is the version needed for the
+proof that `L (f ⍟ g) = L f * L g`. -/
 lemma term_convolution (f g : ℕ → ℂ) (s : ℂ) (n : ℕ) :
```

Comment:
How about
````
let m : ℕ × ℕ → ℕ := fun p ↦ p.1 * p.2
  rcases eq_or_ne n 0 with rfl | hn
  · -- show that both sides vanish when `n = 0`
    refine (term_zero ..).trans ?_ 
    -- the right hand sum is over the union below, but in each term, one factor is always zero
    have hS : m ⁻¹' {0} = {0} ×ˢ univ ∪ univ ×ˢ {0} := by
      ext
      simp only [mem_preimage, mem_singleton_iff, _root_.mul_eq_zero, mem_union, mem_prod, mem_univ,
        and_true, true_and, m]
    convert tsum_zero.symm using 3 with ⟨⟨a, b⟩, hab⟩
    simp only [hS] at hab
    rcases hab with ⟨rfl, -⟩ | ⟨-, rfl⟩ <;> simp only [term_zero, zero_mul, mul_zero]
  · -- now `n > 0`
    have H : m ⁻¹' {n} = n.divisorsAntidiagonal := by
      ext
      simp [m, hn]
    rw [H, Finset.tsum_subtype' n.divisorsAntidiagonal fun x ↦ term f s x.1 * term g s x.2]
    exact term_convolution' ..
````
(sorry, github won't let me make this as a "suggestion")

Reply:
(If you select the range of lines (so that they are highlighted) before clicking on the "Comment" button, and then click on the `±` symbol, the suggestions stuff should work.)

Reply:
(I'm aware of the interface, but github doesn't support making suggestions on line ranges that include one or more deleted lines.)

Reply:
The `convert` is fairly slow, so I do the vanishing of the terms first, but the re-arrangement looks good!

Reply:
(OK; I hadn't noticed that before.)